cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -207,39 +291,182 @@ protected:
+    Status st;
+
+    const size_t nColumnsData  = dim.nCols;
+    const size_t nColumnsPhi   = nColumnsData + 1;
+    const size_t biasTermIndex = nColumnsPhi - 1;
+
+    for (size_t iRow = 0; iRow < nRowsData; ++iRow)
+    {
+        const algorithmFPType * currentX = x + (iRow * nColumnsData);
+        algorithmFPType * phi            = res + (iRow * nColumnsPhi);
+        for (size_t currentTreeIndex = iTree; currentTreeIndex < iTree + nTrees; ++currentTreeIndex)
+        {
+            const gbt::internal::GbtDecisionTree * currentTree = _aTree[currentTreeIndex];
+            st |= gbt::treeshap::treeShap<algorithmFPType, hasUnorderedFeatures, hasAnyMissing>(currentTree, currentX, phi, &_featHelper, condition,
+                                                                                                conditionFeature);
+        }
+
+        if (condition == 0)
+        {
+            // find bias term by leveraging bias = nominal - sum_i phi_i
+            for (int iFeature = 0; iFeature < nColumnsData; ++iFeature)
+            {
+                phi[biasTermIndex] -= phi[iFeature];
+            }
+        }
+    }
+
+    return st;
+}
+
+/**
+ * Helper to predict SHAP contribution interactions
+ * \param[in] iTree index of start tree for the calculation
+ * \param[in] nTrees number of trees in block included in calculation
+ * \param[in] nRowsData number of rows to process
+ * \param[in] nColumnsData number of columns in data, i.e. features
+ *                         note: 1 SHAP value per feature + bias term
+ * \param[in] x pointer to the start of observation data
+ * \param[out] res pointer to the start of memory where results are written to
+ * \param[in] dim DimType helper
+*/
+template <typename algorithmFPType, CpuType cpu>
+template <bool hasUnorderedFeatures, bool hasAnyMissing>
+services::Status PredictRegressionTask<algorithmFPType, cpu>::predictContributionInteractions(size_t iTree, size_t nTrees, size_t nRowsData,
+                                                                                              const algorithmFPType * x,
+                                                                                              const algorithmFPType * nominal, algorithmFPType * res,
+                                                                                              const DimType & dim)
+{
+    Status st;
+    const size_t nColumnsData  = dim.nCols;
+    const size_t nColumnsPhi   = nColumnsData + 1;
+    const size_t biasTermIndex = nColumnsPhi - 1;
+
+    const size_t interactionMatrixSize = nColumnsPhi * nColumnsPhi;
+
+    // Allocate buffer for 3 matrices for algorithmFPType of size (nRowsData, nColumnsData)
+    const size_t elementsInMatrix = nRowsData * nColumnsPhi;
+    const size_t bufferSize       = 3 * sizeof(algorithmFPType) * elementsInMatrix;
+    algorithmFPType * buffer      = static_cast<algorithmFPType *>(daal_calloc(bufferSize));
```

Comment:
Please consider using data management entities for that. It can be just plain `SharedPtr` or `TArray` by choice.